### PR TITLE
Fix cached UI-generated queries in MongoDB

### LINKED ISSUE #25915: Caching not working for some MongoDB questions
**Describe the bug**
Caching not working for some MongoDB questions, and seems to be specific to GUI questions, since similar Native question caches.

**To Reproduce**
1. Create a MongoDB collection (table) like this, and sync the database
```
{ "_id" : ObjectId("63481d83b8fc3a1b5ebf3156"), "item" : "journal", "qty" : 25 }
{ "_id" : ObjectId("63481d83b8fc3a1b5ebf3157"), "item" : "mat", "qty" : 85 }
{ "_id" : ObjectId("63481d83b8fc3a1b5ebf3158"), "item" : "mousepad", "qty" : 25 }
```
2. Admin > Settings > Caching - enable and set Duration=0.0001 and TTL=1000000
3. Question > Mongo > select the table just created and save question
4. Refresh the saved question a couple of times, it will not show the normal cache indicator ("Updated a minute ago" etc) in the bottom right, and the log will have stacktrace about `Error saving query results to cache: backing-map`

<details><summary>Full stacktrace</summary>

```
2022-10-13 18:34:15,050 INFO query-processor.card :: Question's average execution duration is 105.0 ms; using 'magic' TTL of 1.2 days 💾
2022-10-13 18:34:15,159 INFO middleware.cache :: Query took 77.0 ms to run; minimum for cache eligibility is 100.0 µs
2022-10-13 18:34:15,160 INFO middleware.cache :: Caching results for next time for query with hash "64564e93". 💾
2022-10-13 18:34:15,160 ERROR middleware.cache :: Error saving query results to cache: backing-map
java.lang.NoSuchFieldException: backing-map
	at java.base/java.lang.Class.getField(Class.java:1999)
	at taoensso.nippy$fn__25232$fn__25233$fn__25234.invoke(nippy.clj:1209)
	at taoensso.nippy$fn__25232$fn__25233.invoke(nippy.clj:1207)
	at clojure.lang.PersistentVector.reduce(PersistentVector.java:343)
	at clojure.core$reduce.invokeStatic(core.clj:6885)
	at clojure.core$reduce.invoke(core.clj:6868)
	at taoensso.nippy$fn__25232.invokeStatic(nippy.clj:1207)
	at taoensso.nippy$fn__25232.invoke(nippy.clj:1200)
	at taoensso.nippy$fn__25026$G__25021__25033.invoke(nippy.clj:567)
	at taoensso.nippy$fn__25057.invokeStatic(nippy.clj:576)
	at taoensso.nippy$fn__25057.invoke(nippy.clj:569)
	at taoensso.nippy$fn__25044$G__25039__25051.invoke(nippy.clj:568)
	at taoensso.nippy$write_map$fn__25118$fn__25119.invoke(nippy.clj:887)
	at taoensso.nippy$write_map$fn__25118.invoke(nippy.clj:884)
	at clojure.lang.PersistentArrayMap.kvreduce(PersistentArrayMap.java:429)
	at clojure.core$fn__8525.invokeStatic(core.clj:6908)
	at clojure.core$fn__8525.invoke(core.clj:6888)
	at clojure.core.protocols$fn__8257$G__8252__8266.invoke(protocols.clj:175)
	at clojure.core$reduce_kv.invokeStatic(core.clj:6919)
	at clojure.core$reduce_kv.invoke(core.clj:6910)
	at taoensso.nippy$write_map.invokeStatic(nippy.clj:884)
	at taoensso.nippy$write_map.invoke(nippy.clj:866)
	at taoensso.nippy$fn__25222.invokeStatic(nippy.clj:1172)
	at taoensso.nippy$fn__25222.invoke(nippy.clj:1172)
	at taoensso.nippy$fn__25026$G__25021__25033.invoke(nippy.clj:567)
	at taoensso.nippy$fn__25057.invokeStatic(nippy.clj:576)
	at taoensso.nippy$fn__25057.invoke(nippy.clj:569)
	at taoensso.nippy$fn__25044$G__25039__25051.invoke(nippy.clj:568)
	at taoensso.nippy$write_vec$fn__25089$fn__25090.invoke(nippy.clj:751)
	at taoensso.nippy$write_vec$fn__25089.invoke(nippy.clj:751)
	at clojure.lang.PersistentVector.reduce(PersistentVector.java:343)
	at clojure.core$reduce.invokeStatic(core.clj:6885)
	at clojure.core$reduce.invoke(core.clj:6868)
	at taoensso.nippy$write_vec.invokeStatic(nippy.clj:751)
	at taoensso.nippy$write_vec.invoke(nippy.clj:729)
	at taoensso.nippy$fn__25218.invokeStatic(nippy.clj:1170)
	at taoensso.nippy$fn__25218.invoke(nippy.clj:1170)
	at taoensso.nippy$fn__25026$G__25021__25033.invoke(nippy.clj:567)
	at taoensso.nippy$fn__25057.invokeStatic(nippy.clj:576)
	at taoensso.nippy$fn__25057.invoke(nippy.clj:569)
	at taoensso.nippy$fn__25044$G__25039__25051.invoke(nippy.clj:568)
	at taoensso.nippy$write_map$fn__25118$fn__25119.invoke(nippy.clj:887)
	at taoensso.nippy$write_map$fn__25118.invoke(nippy.clj:884)
	at clojure.lang.PersistentArrayMap.kvreduce(PersistentArrayMap.java:429)
	at clojure.core$fn__8525.invokeStatic(core.clj:6908)
	at clojure.core$fn__8525.invoke(core.clj:6888)
	at clojure.core.protocols$fn__8257$G__8252__8266.invoke(protocols.clj:175)
	at clojure.core$reduce_kv.invokeStatic(core.clj:6919)
	at clojure.core$reduce_kv.invoke(core.clj:6910)
	at taoensso.nippy$write_map.invokeStatic(nippy.clj:884)
	at taoensso.nippy$write_map.invoke(nippy.clj:866)
	at taoensso.nippy$fn__25222.invokeStatic(nippy.clj:1172)
	at taoensso.nippy$fn__25222.invoke(nippy.clj:1172)
	at taoensso.nippy$fn__25026$G__25021__25033.invoke(nippy.clj:567)
	at taoensso.nippy$fn__25057.invokeStatic(nippy.clj:576)
	at taoensso.nippy$fn__25057.invoke(nippy.clj:569)
	at taoensso.nippy$fn__25044$G__25039__25051.invoke(nippy.clj:568)
	at taoensso.nippy$write_map$fn__25118$fn__25119.invoke(nippy.clj:887)
	at taoensso.nippy$write_map$fn__25118.invoke(nippy.clj:884)
	at clojure.lang.PersistentArrayMap.kvreduce(PersistentArrayMap.java:429)
	at clojure.core$fn__8525.invokeStatic(core.clj:6908)
	at clojure.core$fn__8525.invoke(core.clj:6888)
	at clojure.core.protocols$fn__8257$G__8252__8266.invoke(protocols.clj:175)
	at clojure.core$reduce_kv.invokeStatic(core.clj:6919)
	at clojure.core$reduce_kv.invoke(core.clj:6910)
	at taoensso.nippy$write_map.invokeStatic(nippy.clj:884)
	at taoensso.nippy$write_map.invoke(nippy.clj:866)
	at taoensso.nippy$fn__25222.invokeStatic(nippy.clj:1172)
	at taoensso.nippy$fn__25222.invoke(nippy.clj:1172)
	at taoensso.nippy$fn__25026$G__25021__25033.invoke(nippy.clj:567)
	at taoensso.nippy$fn__25057.invokeStatic(nippy.clj:576)
	at taoensso.nippy$fn__25057.invoke(nippy.clj:569)
	at taoensso.nippy$fn__25044$G__25039__25051.invoke(nippy.clj:568)
	at taoensso.nippy$write_map$fn__25118$fn__25119.invoke(nippy.clj:887)
	at taoensso.nippy$write_map$fn__25118.invoke(nippy.clj:884)
	at clojure.lang.PersistentArrayMap.kvreduce(PersistentArrayMap.java:429)
	at clojure.core$fn__8525.invokeStatic(core.clj:6908)
	at clojure.core$fn__8525.invoke(core.clj:6888)
	at clojure.core.protocols$fn__8257$G__8252__8266.invoke(protocols.clj:175)
	at clojure.core$reduce_kv.invokeStatic(core.clj:6919)
	at clojure.core$reduce_kv.invoke(core.clj:6910)
	at taoensso.nippy$write_map.invokeStatic(nippy.clj:884)
	at taoensso.nippy$write_map.invoke(nippy.clj:866)
	at taoensso.nippy$fn__25222.invokeStatic(nippy.clj:1172)
	at taoensso.nippy$fn__25222.invoke(nippy.clj:1172)
	at taoensso.nippy$fn__25026$G__25021__25033.invoke(nippy.clj:567)
	at taoensso.nippy$fn__25057.invokeStatic(nippy.clj:576)
	at taoensso.nippy$fn__25057.invoke(nippy.clj:569)
	at taoensso.nippy$fn__25044$G__25039__25051.invoke(nippy.clj:568)
	at taoensso.nippy$freeze_to_out_BANG_.invokeStatic(nippy.clj:1011)
	at taoensso.nippy$freeze_to_out_BANG_.invoke(nippy.clj:1008)
	at metabase.query_processor.middleware.cache.impl$freeze_BANG_.invokeStatic(impl.clj:36)
	at metabase.query_processor.middleware.cache.impl$freeze_BANG_.invoke(impl.clj:33)
	at metabase.query_processor.middleware.cache.impl$do_with_serialization$in_STAR___49015$fn__49016.invoke(impl.clj:71)
	at metabase.query_processor.middleware.cache.impl$do_with_serialization$in_STAR___49015.invoke(impl.clj:70)
	at metabase.query_processor.middleware.cache$add_object_to_cache_BANG_.invokeStatic(cache.clj:64)
	at metabase.query_processor.middleware.cache$add_object_to_cache_BANG_.invoke(cache.clj:60)
	at metabase.query_processor.middleware.cache$save_results_xform$fn__52571.invoke(cache.clj:103)
	at clojure.core$transduce.invokeStatic(core.clj:6948)
	at clojure.core$transduce.invokeStatic(core.clj:6942)
	at clojure.core$transduce.invoke(core.clj:6933)
	at metabase.query_processor.context.default$default_reducef$fn__39174.invoke(default.clj:57)
	at metabase.query_processor.context.default$default_reducef.invokeStatic(default.clj:56)
	at metabase.query_processor.context.default$default_reducef.invoke(default.clj:48)
	at metabase.query_processor.middleware.cache$run_query_with_cache$reducef_SINGLEQUOTE___52607$fn__52608.invoke(cache.clj:192)
	at metabase.query_processor.middleware.cache.impl$do_with_serialization.invokeStatic(impl.clj:68)
	at metabase.query_processor.middleware.cache.impl$do_with_serialization.invoke(impl.clj:39)
	at metabase.query_processor.middleware.cache.impl$do_with_serialization.invokeStatic(impl.clj:58)
	at metabase.query_processor.middleware.cache.impl$do_with_serialization.invoke(impl.clj:39)
	at metabase.query_processor.middleware.cache$run_query_with_cache$reducef_SINGLEQUOTE___52607.invoke(cache.clj:188)
	at metabase.query_processor.context$reducef.invokeStatic(context.clj:69)
	at metabase.query_processor.context$reducef.invoke(context.clj:62)
	at metabase.query_processor.context.default$default_runf$respond_STAR___39178.invoke(default.clj:68)
	at metabase.driver.mongo.execute$reduce_results.invokeStatic(execute.clj:138)
	at metabase.driver.mongo.execute$reduce_results.invoke(execute.clj:131)
	at metabase.driver.mongo.execute$execute_reducible_query.invokeStatic(execute.clj:169)
	at metabase.driver.mongo.execute$execute_reducible_query.invoke(execute.clj:157)
	at metabase.driver.mongo$fn__96248$f__94573__auto____96249.invoke(mongo.clj:238)
	at metabase.driver.mongo.util$do_with_mongo_connection$fn__94564.invoke(util.clj:226)
	at metabase.util.ssh$do_with_ssh_tunnel.invokeStatic(ssh.clj:134)
	at metabase.util.ssh$do_with_ssh_tunnel.invoke(ssh.clj:123)
	at metabase.driver.mongo.util$do_with_mongo_connection.invokeStatic(util.clj:220)
	at metabase.driver.mongo.util$do_with_mongo_connection.invoke(util.clj:215)
	at metabase.driver.mongo$fn__96248.invokeStatic(mongo.clj:237)
	at metabase.driver.mongo$fn__96248.invoke(mongo.clj:235)
	at clojure.lang.MultiFn.invoke(MultiFn.java:244)
	at metabase.query_processor.context$executef.invokeStatic(context.clj:59)
	at metabase.query_processor.context$executef.invoke(context.clj:48)
	at metabase.query_processor.context.default$default_runf.invokeStatic(default.clj:67)
	at metabase.query_processor.context.default$default_runf.invoke(default.clj:65)
	at metabase.query_processor.context$runf.invokeStatic(context.clj:45)
	at metabase.query_processor.context$runf.invoke(context.clj:39)
	at metabase.query_processor.reducible$identity_qp.invokeStatic(reducible.clj:12)
	at metabase.query_processor.reducible$identity_qp.invoke(reducible.clj:9)
	at metabase.query_processor.middleware.cache$run_query_with_cache.invokeStatic(cache.clj:193)
	at metabase.query_processor.middleware.cache$run_query_with_cache.invoke(cache.clj:178)
	at metabase.query_processor.middleware.cache$maybe_return_cached_results$maybe_return_cached_results_STAR___52619.invoke(cache.clj:219)
	at metabase.query_processor.middleware.permissions$check_query_permissions$fn__48140.invoke(permissions.clj:109)
	at metabase.query_processor.middleware.mbql_to_native$mbql__GT_native$fn__51564.invoke(mbql_to_native.clj:23)
	at metabase.query_processor$fn__54095$combined_post_process__54100$combined_post_process_STAR___54101.invoke(query_processor.clj:212)
	at metabase.query_processor$fn__54095$combined_pre_process__54096$combined_pre_process_STAR___54097.invoke(query_processor.clj:209)
	at metabase.query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__52540$fn__52545.invoke(resolve_database_and_driver.clj:35)
	at metabase.driver$do_with_driver.invokeStatic(driver.clj:75)
	at metabase.driver$do_with_driver.invoke(driver.clj:71)
	at metabase.query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__52540.invoke(resolve_database_and_driver.clj:34)
	at metabase.query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__48575.invoke(fetch_source_query.clj:353)
	at metabase.query_processor.middleware.store$initialize_store$fn__48763$fn__48764.invoke(store.clj:11)
	at metabase.query_processor.store$do_with_store.invokeStatic(store.clj:44)
	at metabase.query_processor.store$do_with_store.invoke(store.clj:38)
	at metabase.query_processor.middleware.store$initialize_store$fn__48763.invoke(store.clj:10)
	at metabase.query_processor.middleware.normalize_query$normalize$fn__52812.invoke(normalize_query.clj:22)
	at metabase.query_processor.middleware.constraints$add_default_userland_constraints$fn__49845.invoke(constraints.clj:53)
	at metabase.query_processor.middleware.process_userland_query$process_userland_query$fn__52751.invoke(process_userland_query.clj:145)
	at metabase.query_processor.middleware.catch_exceptions$catch_exceptions$fn__53123.invoke(catch_exceptions.clj:167)
	at metabase.query_processor.reducible$async_qp$qp_STAR___44978$thunk__44980.invoke(reducible.clj:100)
	at metabase.query_processor.reducible$async_qp$qp_STAR___44978$fn__44982.invoke(reducible.clj:105)
	at clojure.core$binding_conveyor_fn$fn__5823.invoke(core.clj:2047)
	at clojure.lang.AFn.call(AFn.java:18)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
2022-10-13 18:34:15,168 DEBUG middleware.log :: POST /api/card/1123/query 202 [ASYNC: completed] 121.1 ms (17 DB calls) App DB connections: 1/15 Jetty threads: 3/50 (2 idle, 0 queued) (117 total active threads) Queries in flight: 0 (0 queued)
```

</details>

**Information about your Metabase Installation:**
Tested 0.44.4

**Additional context**
Seems very related to #11995 and #19501

### PR DESCRIPTION
**Pull request by john-metabase**
This fixes query caching on MongoDB (resolves #25915) for most UI-generated queries.

[EDIT:] Updated to handle this correctly via `extend-freeze`/`extend-thaw` instead of JSON. 🤦

### CODE REVIEW

**Comment by qnkhuat** [comment]
should we add some test for this?

**Review by metamben — COMMENTED** [review]
- Since `ordered-map` is not mongo-specific, I guess this should not be in a mongo namespace.
- A test reproducing the bug in #25915 should be added.

**Comment by john-metabase** [comment]
Yeah, agreed on the tests, I went to add a test to this today but didn't quite get there. I was hoping I could just turn on the `:mongo` driver for a cached query test but it doesn't look like one exists. Does @qnkhuat or anyone know if there is already a test like that? If there is, it must be skipping the particular failure case for MongoDB.

@metamben I will move it to `metabase.query-processor.middleware.cache` instead.

**Review by metamben — APPROVED** [review]
👍

File: bin/build-drivers/deps.edn [hunk 1]
```diff
@@ -11,7 +11,6 @@
   io.forward/yaml                 {:mvn/version "1.0.9"} ; Don't upgrade yet, new version doesn't support Java 8 (see https://github.com/owainlewis/yaml/issues/37)
   io.github.clojure/tools.build   {:git/tag "v0.7.4" :git/sha "ac442da"}
   org.clojure/tools.deps.alpha    {:mvn/version "0.12.985"}
-  org.flatland/ordered            {:mvn/version "1.5.9"} ; used by io.forward/yaml -- need the newer version
```
Comment (john-metabase) [inline bin/build-drivers/deps.edn:14]: @camsaul I wanted to check with you on this one. The version pulled in by this `deps.edn` causes a failure in `bin/build-drivers` tests. Removing this solves it, but so would just adding the same version that is in the root `deps.edn`. I'm not clear on why there is an explicit dependency on this lib in this project. Without removing or updating flatland here, the test throws an error like:

```
Step failed: Syntax error compiling at (metabase/query_processor/middleware/cache/impl.clj:44:20).
{:via
 [{:type clojure.lang.Compiler$CompilerException,
   :message "Syntax error compiling at (metabase/query_processor/middleware/cache/impl.clj:44:20).",
   :data
   #:clojure.error{:phase :compile-syntax-check,
                   :line 44,
                   :column 20,
                   :source "metabase/query_processor/middleware/cache/impl.clj"},
   :at [clojure.lang.Compiler analyze "Compiler.java" 6825]}
  {:type java.lang.RuntimeException,
   :message "No such var: ordered-map/ordered-map-reader",
   :at [clojure.lang.Util runtimeException "Util.java" 221]}],
 :trace
 ...
 ```

Comment (camsaul) [reply bin/build-drivers/deps.edn:14]: > I'm not clear on why there is an explicit dependency on this lib in this project.

```clj
; used by io.forward/yaml -- need the newer version
```

that's why

Comment (camsaul) [reply bin/build-drivers/deps.edn:14]: bumping it to the latest version is good, if it works now if you remove it then that's fine too

**Review by camsaul — APPROVED** [review]

**Pull request merged**